src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/CommitManager.cs

```diff
@@ -0,0 +1,196 @@
+        /// </summary>
+        private bool IsCommitCharacter(char typeChar, ImmutableArray<CharacterSetModificationRule> rules)
+        {
+            // Tab, Enter and Null (call invoke commit) are always a commit character
+            if (typeChar == '\t' || typeChar == '\n' || typeChar == '\0')
+            {
+                return true;
+            }
+
+            foreach (var rule in rules)
+            {
+                switch (rule.Kind)
+                {
+                    case CharacterSetModificationKind.Add:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return true;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Remove:
+                        if (rule.Characters.Contains(typeChar))
+                        {
+                            return false;
+                        }
+
+                        break;
+
+                    case CharacterSetModificationKind.Replace:
+                        return rule.Characters.Contains(typeChar);
+                }
+            }
+
+            return s_commitChars.Contains(typeChar);
+        }
+
+        private AsyncCompletionData.CommitBehavior CustomCommit(
+            Document document,
+            IFeaturesCustomCommitCompletionProvider provider,
+            ITextView view,
+            ITextBuffer subjectBuffer,
+            RoslynCompletionItem roslynItem,
+            ITrackingSpan applicableSpan,
+            char? commitCharacter,
+            CancellationToken cancellationToken)
+        {
+            bool includesCommitCharacter;
+            if (!subjectBuffer.CheckEditAccess())
+            {
+                // We are on the wrong thread.
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            if (subjectBuffer.EditInProgress)
+            {
+                return AsyncCompletionData.CommitBehavior.None;
+            }
+
+            using (var edit = subjectBuffer.CreateEdit())
+            {
+                var currentViewSpan = applicableSpan.GetSpan(applicableSpan.TextBuffer.CurrentSnapshot);
+                var spans = view.BufferGraph.MapDownToBuffer(currentViewSpan, SpanTrackingMode.EdgeInclusive, subjectBuffer);
+
+                // There can be no spans in case of projections but must be less than 2 spans.
+                // 1 span is a regular case.
+                // 0 spans means the a span is in anoter projection.
+                // We do not expect more than 1 span, and do not support this case.
+                System.Diagnostics.Debug.Assert(spans.Count < 2);
+                if (spans.Any())
+                {
+                    var currentBufferSpan = spans[0].TranslateTo(subjectBuffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive);
+                    edit.Delete(currentBufferSpan);
```

Comment:
needs comment explainint why it is edge-inclusive.

Reply:
removed this area. borrowed the approach from the old completion

---
In reply to: [222520319](https://github.com/dotnet/roslyn/pull/29016#discussion_r222520319) [](ancestors = 222520319)